cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,288 @@
+			return nil, err
+		}
+	}
+
+	fc := &FirewallController{
+		kubeClient:      kubeClient,
+		client:          client,
+		firewallManager: fwManager,
+		fwCache: &firewallCache{
+			mu: new(sync.RWMutex),
+		},
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				fc.onServiceChange(ctx)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				fc.onServiceChange(ctx)
+			},
+			DeleteFunc: func(obj interface{}) {
+				fc.onServiceChange(ctx)
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc, nil
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}) {
+	ctx, cancel := context.WithTimeout(context.Background(), serviceSyncPeriod)
+	defer cancel()
+
+	wait.Until(func() {
+		targetFirewallState := fc.fwCache.firewall.state
+		fm := *fc.firewallManager
+
+		currentFirewallState, err := fm.Get(context.Background())
+		if err != nil {
+			klog.Errorf("failed to get worker firewall: %s", err)
+		}
+
+		if &currentFirewallState != nil {
+			if cmp.Equal(currentFirewallState, targetFirewallState) {
+				return
+			}
+		}
+		err = fc.onServiceChange(ctx)
+		if err != nil {
+			klog.Errorf("Failed to reconcile worker firewall: %s", err)
+		}
+
+	}, 5*time.Minute, stopCh)
+}
+
+// Get returns the current CCM worker firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	if notNil(fm) {
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fm.fwCache.firewall.state.ID)
+		if err != nil {
+			klog.Errorf("failed to retrieve firewall by ID from local cache")
+		}
+		if resp != nil && resp.StatusCode == 404 {
+			klog.Infof("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	firewallList, err := allFirewallList(ctx, fm.client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve firewall from DO firewall API: %s", err)
+	}
+	formattedWorkerFirewallName := fmt.Sprintf(firewallWorkerCCMNameFormat, fm.workerFirewallName)
+	for _, fw := range firewallList {
+		if fw.Name == formattedWorkerFirewallName {
+			return &fw, nil
+		}
+	}
+	return nil, nil
+}
+
+func notNil(fm *firewallManagerOp) bool {
+	if fm.fwCache == nil || fm.fwCache.firewall == nil || fm.fwCache.firewall.state == nil {
+		return false
+	}
+	return true
+}
+
+// Set applies the given inbound rules to the CCM worker firewall when the current rules and target rules differ.
+func (fm *firewallManagerOp) Set(ctx context.Context, svcInboundRules []godo.InboundRule) error {
+	fm.fwCache.mu.RLock()
```

Comment:
This (read) lock only needs to encompass the `cmp.Equal` call. In particular, we do not need to `defer` until the end of the function.

Suggestion:

```go
fm.fwCache.mu.RLock()
isEqual := cmp.Equal(targetFirewall.InboundRules, svcInboundRules)
fm.fwCache.mu.RUnlock()

if !isEqual {
    [...]